### PR TITLE
Fix target counting in strings char-parallel replace

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Replace `thrust::count_if` call across int64 characters to use a custom kernel instead.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by davidwendt** [comment]
No change in performance according to the benchmark
```
Benchmark                                                      Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------
StringReplace/scalar/4096/32/manual_time                    -0.0002         -0.0008             0             0             0             0
StringReplace/scalar/4096/64/manual_time                    +0.0068         +0.0054             0             0             0             0
StringReplace/scalar/4096/128/manual_time                   +0.0081         +0.0067             0             0             0             0
StringReplace/scalar/4096/256/manual_time                   +0.0065         +0.0056             0             0             0             0
StringReplace/scalar/4096/512/manual_time                   -0.0188         -0.0181             0             0             0             0
StringReplace/scalar/4096/1024/manual_time                  -0.0043         -0.0043             0             0             0             0
StringReplace/scalar/4096/2048/manual_time                  -0.0013         -0.0010             1             1             1             1
StringReplace/scalar/4096/4096/manual_time                  -0.0001         -0.0001             1             1             1             1
StringReplace/scalar/4096/8192/manual_time                  +0.0068         +0.0068             2             2             2             2
StringReplace/scalar/32768/32/manual_time                   +0.0067         +0.0068             0             0             0             0
StringReplace/scalar/32768/64/manual_time                   +0.0162         +0.0155             0             0             0             0
StringReplace/scalar/32768/128/manual_time                  +0.0103         +0.0099             0             0             0             0
StringReplace/scalar/32768/256/manual_time                  +0.0140         +0.0135             0             0             0             0
StringReplace/scalar/32768/512/manual_time                  +0.0099         +0.0097             1             1             1             1
StringReplace/scalar/32768/1024/manual_time                 +0.0053         +0.0054             1             1             1             1
StringReplace/scalar/32768/2048/manual_time                 +0.0166         +0.0164             2             2             2             2
StringReplace/scalar/32768/4096/manual_time                 +0.0154         +0.0153             3             3             3             3
StringReplace/scalar/32768/8192/manual_time                 +0.0166         +0.0166             6             6             6             6
StringReplace/scalar/262144/32/manual_time                  +0.0052         +0.0047             0             0             0             0
StringReplace/scalar/262144/64/manual_time                  +0.0091         +0.0080             0             0             0             0
StringReplace/scalar/262144/128/manual_time                 +0.0172         +0.0165             1             1             1             1
StringReplace/scalar/262144/256/manual_time                 +0.0064         +0.0064             2             2             2             2
StringReplace/scalar/262144/512/manual_time                 +0.0121         +0.0121             4             4             4             4
StringReplace/scalar/262144/1024/manual_time                +0.0068         +0.0068             8             8             8             8
StringReplace/scalar/262144/2048/manual_time                -0.0083         -0.0083            22            21            22            21
StringReplace/scalar/262144/4096/manual_time                -0.0056         -0.0055            47            46            47            46
StringReplace/scalar/2097152/32/manual_time                 +0.0104         +0.0099             1             1             1             1
StringReplace/scalar/2097152/64/manual_time                 +0.0102         +0.0100             2             2             2             2
StringReplace/scalar/2097152/128/manual_time                +0.0122         +0.0120             4             4             4             4
StringReplace/scalar/2097152/256/manual_time                +0.0066         +0.0066            13            13            13            13
StringReplace/scalar/2097152/512/manual_time                +0.0161         +0.0161            26            27            26            27
StringReplace/scalar/16777216/32/manual_time                +0.0086         +0.0086             5             5             5             5
StringReplace/scalar/16777216/64/manual_time                +0.0140         +0.0140            13            13            13            13

```

**Comment by vuule** [comment]
what's the issue with count_if?

**Comment by davidwendt** [comment]
> what's the issue with count_if?

The `thrust::count_if` available to us only supports a max-int32 range of values. That is, `thrust::distance(begin,end) < max-int32`
The int64 ranged version is a separate implementation in thrust and which is one of the few we currently compile out to save space in libcudf.so.

File: cpp/tests/large_strings/replace_tests.cpp [hunk 1]
```diff
@@ -65,3 +66,22 @@ TEST_F(StringsManyTest, Replace)
...
+  auto const expected   = this->long_column();
+  auto const view       = cudf::column_view(expected);
+  auto const multiplier = 10;
+  auto const separator  = cudf::string_scalar("|");
+  auto const input      = cudf::strings::concatenate(
+    cudf::table_view(std::vector<cudf::column_view>(multiplier, view)), separator);
+
+  auto const input_view = cudf::strings_column_view(input->view());
+  auto const target     = cudf::string_scalar("3");  // fake the actual replace;
+  auto const repl       = cudf::string_scalar("3");  // logic still builds the output
```
Comment (vuule) [inline cpp/tests/large_strings/replace_tests.cpp:81]: are we certain this will always be the case?

Comment (davidwendt) [reply cpp/tests/large_strings/replace_tests.cpp:81]: So, I don't see a future where we copy the target/repl from the device or launch a single device thread to check for these being equal.

File: cpp/src/strings/replace/replace.cu [hunk 2]
```diff
@@ -238,6 +238,31 @@ struct replace_parallel_chars_fn {
...
+
+  int64_t count = 0;
+  // each thread processes multiple bytes
+  for (auto i = byte_idx; (i < (byte_idx + bytes_per_thread)) && (i < chars_bytes); ++i) {
+    count += fn.has_target(i);
+  }
+  auto const total = block_reduce(temp_storage).Reduce(count, cub::Sum());
+
+  if ((lane_idx == 0) && (total > 0)) {
+    cuda::atomic_ref<int64_t, cuda::thread_scope_block> ref{*d_output};
```
Comment (vuule) [inline cpp/src/strings/replace/replace.cu:261]: The use of `thread_scope_block` here is surprising to me. Aren't threads from different blocks updating the same output here? I'd expect that `thread_scope_device` is required.
I'm probably missing something.

Comment (davidwendt) [reply cpp/src/strings/replace/replace.cu]: Sounds right to me. Fixed in https://github.com/rapidsai/cudf/pull/16017/commits/530de96a49e72c338f33c999b2cb2cfa802b94d8

**Review by vuule — APPROVED** [review]

**Review by srinivasyadav18 — APPROVED** [review]
LGTM!

**Pull request merged**